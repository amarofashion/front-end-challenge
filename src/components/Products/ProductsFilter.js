import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  isOnSale: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

const ProductsFilter = ({ isOnSale, onClick }) => {
  return (
    <div className="App__filter">
      <button
        type="button"
        className="App__button App__button--small"
        onClick={onClick}
      >
        Filtrar Promoções -
        {' '}
        {isOnSale ? 'Sim' : 'Não'}
      </button>
    </div>
  );
};

ProductsFilter.propTypes = propTypes;

export default ProductsFilter;
