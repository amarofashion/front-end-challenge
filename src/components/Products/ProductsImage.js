import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  name: PropTypes.string,
  color: PropTypes.string,
  image: PropTypes.string,
  isLarge: PropTypes.bool.isRequired,
};

const defaultProps = {
  name: '',
  color: '',
  image: '',
};

const ProductsImage = ({ name, color, image, isLarge }) => {
  const nameLowerCase = name.toLowerCase();
  const colorLowerCase = color.toLowerCase();

  return (
    <div
      className={`App__products__picture ${
        isLarge ? 'App__products__picture--large' : ''
      }`}
    >
      <picture>
        <img
          src={image || 'https://via.placeholder.com/470x594'}
          className="App__products__photo"
          alt={`${nameLowerCase} - ${colorLowerCase}`}
        />
      </picture>
    </div>
  );
};

ProductsImage.propTypes = propTypes;
ProductsImage.defaultProps = defaultProps;

export default ProductsImage;
