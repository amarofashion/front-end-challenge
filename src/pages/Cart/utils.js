import { translateToNumber } from '../../utils/number';

export function getPrices(list) {
  const result = {
    subTotal: 0,
    totalDiscount: 0,
    total: 0,
  };

  list.forEach(({ onSale, actualPrice, regularPrice, quantity }) => {
    const actualPriceValue = translateToNumber(actualPrice) * quantity;
    const regularPriceValue = translateToNumber(regularPrice) * quantity;

    if (onSale) {
      result.totalDiscount += regularPriceValue - actualPriceValue;
      result.total += actualPriceValue;
    } else {
      result.total += regularPriceValue;
    }

    result.subTotal += regularPriceValue;
  });

  return result;
}
