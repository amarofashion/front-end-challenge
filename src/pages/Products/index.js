import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { getProducts, addProductToCart } from '../../store/actions';
import Product from '../../components/Product';
import DetailedProduct from '../../components/DetailedProduct';
import Modal from '../../components/Modal';
import Icon from '../../components/Icon';
import { StyledWrapper } from './style';

function Products() {
  const dispatch = useDispatch();

  const isSaleSelected = useSelector(state => state.productsFilters.sale);

  let list = useSelector(state => state.products);
  if (isSaleSelected) list = list.filter(item => item.onSale);

  const [selected, setSelected] = useState({});
  const [filter, setFilter] = useState(true);

  const hasSelected = Object.keys(selected).length > 0;

  const selectProduct = product => {
    dispatch({ type: 'OPEN_MODAL' });
    setSelected(product);
  };

  const updateFilter = () => {
    dispatch({ type: 'UPDATE_SALE_FILTER' });
    setFilter(!filter);
  };

  const addToCart = ({ product, selectedSize }) => {
    dispatch(addProductToCart({ product, selectedSize }));
    dispatch({ type: 'CLOSE_MODAL' });
  };

  useEffect(() => {
    if (list.length === 0) dispatch(getProducts());
  });

  return (
    <StyledWrapper>
      {hasSelected && (
        <Modal>
          <DetailedProduct
            {...selected}
            callback={selectedSize =>
              addToCart({ product: selected, selectedSize })
            }
          />
        </Modal>
      )}

      <div className='order'>
        <button type='button' onClick={() => updateFilter()}>
          only <span>sale</span>
          <Icon name={filter ? 'down' : 'delete'} size='1em' />
        </button>
      </div>

      <ul>
        {list.map((product, index) => (
          <li key={`product-${index}`} data-cypress='products-item'>
            <Product {...product} callback={() => selectProduct(product)} />
          </li>
        ))}
      </ul>
    </StyledWrapper>
  );
}

export default Products;
