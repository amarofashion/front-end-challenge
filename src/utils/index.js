const replaceValue = value => {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^\w]+/g, '-');
};

const friendlyURL = (name = '', color = '', style = '') => {
  return replaceValue(name)
    .concat(`-${replaceValue(color)}`)
    .concat(`-${style}`);
};

export default friendlyURL;
