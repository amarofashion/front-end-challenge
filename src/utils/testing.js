import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { initialState } from '../store/reducers';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

// eslint-disable-next-line react/prop-types
export const WithStore = ({ children }) => (
  <Provider store={store}>{children}</Provider>
);

// eslint-disable-next-line react/prop-types
export const WithRouter = ({ children }) => <Router>{children}</Router>;
