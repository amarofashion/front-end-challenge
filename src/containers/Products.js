import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import withLoader from '../components/withLoader';

import { setProducts } from '../actions/store';

import ProductsFilter from '../components/Products/ProductsFilter';
import ProductsList from '../components/Products/ProductsList';

import productsJSON from '../data/products.json';

const propTypes = {
  products: PropTypes.instanceOf(Array).isRequired,
  setProducts: PropTypes.func.isRequired,
};

class Products extends PureComponent {
  constructor() {
    super();

    this.state = {
      isOnSale: false,
    };
  }

  componentDidMount() {
    const { products } = this.props;

    if (!products.length) {
      this.fetchAllProducts();
    }
  }

  fetchAllProducts = () => {
    this.addProducts(productsJSON);
  };

  addProducts = data => {
    const { setProducts } = this.props;

    setProducts({
      products: data,
    });
  };

  handleOnSale = () => {
    this.setState(prevState => ({
      isOnSale: !prevState.isOnSale,
    }));
  };

  render() {
    const { products } = this.props;
    const { isOnSale } = this.state;

    return (
      <React.Fragment>
        <ProductsFilter isOnSale={isOnSale} onClick={this.handleOnSale} />

        <ProductsList
          products={products}
          isOnSale={isOnSale}
          isLoading={false}
        />
      </React.Fragment>
    );
  }
}

Products.propTypes = propTypes;

const mapStateToProps = state => {
  return {
    ...state.store,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    setProducts: data => {
      dispatch(setProducts(data));
    },
  };
};

export default withLoader(
  withRouter(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(Products)
  )
);
