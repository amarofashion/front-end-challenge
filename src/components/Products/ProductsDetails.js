import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { setBasket } from '../../actions/products';

import ProductsTitle from './ProductsTitle';
import ProductsImage from './ProductsImage';
import ProductsPrices from './ProductsPrices';
import ProductsSizes from './ProductsSizes';

const propTypes = {
  products: PropTypes.instanceOf(Array),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }),
  setBasket: PropTypes.func,
};

const defaultProps = {
  products: [],
  history: {
    location: {
      name: '',
    },
  },
  setBasket: () => {},
};

class ProductsDetails extends PureComponent {
  constructor() {
    super();

    this.state = {
      productName: '',
      productSize: '',
      productAmount: 1,
    };
  }

  componentDidMount() {
    window.scrollTo(0, 0);
  }

  getProduct = (style, code_color) => {
    const { products } = this.props;

    return (
      products.find(
        product => product.style === style && product.code_color === code_color
      ) || {}
    );
  };

  handleBasketAmount = event => {
    this.setState({ productAmount: event.target.value });
  };

  handleAdd = () => {
    const { productName, productSize, productAmount } = this.state;
    const { setBasket } = this.props;

    setBasket(productName, productSize, productAmount);
  };

  handleGoBack = () => {
    const { history } = this.props;
    history.goBack();
  };

  render() {
    const { products } = this.props;
    const { history } = this.props;

    if (!products.length) {
      history.goBack();
    }

    const {
      history: {
        location: {
          state: { style, code_color },
        },
      },
    } = this.props;

    const {
      name,
      actual_price,
      color,
      discount_percentage,
      image,
      installments,
      on_sale,
      regular_price,
      sizes,
    } = this.getProduct(style, code_color);

    return (
      <div className="App__products__item App__products__item--large">
        <ProductsTitle name={name} color={color} isTitle />

        <div className="App__products__main">
          <ProductsImage name={name} color={color} image={image} isLarge />

          <div className="App__products__info">
            <ProductsPrices
              on_sale={on_sale}
              regular_price={regular_price}
              actual_price={actual_price}
              discount_percentage={discount_percentage}
            />

            <p className="App__products__details">{installments}</p>

            <ProductsSizes sizes={sizes} />

            <input
              type="number"
              id="productAmount"
              defaultValue="1"
              className="App__input App__input--amount"
              onChange={this.handleBasketAmount}
            />
            <button
              type="button"
              className="App__button App__button--more-detail"
              onClick={() => this.handleAdd(name, 1, 2)}
            >
              {'Adicionar'}
            </button>
          </div>
        </div>

        <button
          type="button"
          className="App__button App__button--more-detail"
          onClick={this.handleGoBack}
        >
          {'Voltar'}
        </button>
      </div>
    );
  }
}

ProductsDetails.propTypes = propTypes;
ProductsDetails.defaultProps = defaultProps;

const mapStateToProps = state => {
  return {
    ...state.store,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    setBasket: (name, size, qtd) => {
      dispatch(setBasket(name, size, qtd));
    },
  };
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ProductsDetails)
);
