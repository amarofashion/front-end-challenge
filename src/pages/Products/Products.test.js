import React from 'react';
import { mount } from 'enzyme';

import Products from '.';
import { WithStore, WithRouter } from '../../utils/testing';

describe('<Products />', () => {
  let wrapper;

  describe('with default props', () => {
    beforeAll(() => {
      wrapper = mount(
        <WithStore>
          <WithRouter>
            <Products />
          </WithRouter>
        </WithStore>
      );
    });

    it('shoud match snapshot with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
