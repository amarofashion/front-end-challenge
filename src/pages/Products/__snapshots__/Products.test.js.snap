// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> with default props shoud match snapshot with default props 1`] = `
<WithStore>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <WithRouter>
      <BrowserRouter>
        <Router
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
        >
          <Products>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htoDjs",
                      "isStatic": false,
                      "lastClassName": "eVHVmW",
                      "rules": Array [
                        "
  ul {
    margin-left: -15px;
    margin-right: -15px;
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;

    & > li:last-child,
    & > li:nth-last-child(2),
    & > li:nth-last-child(3) {
      flex-grow: 0;
    }

    li {
      width: calc(100% - 20px);
      margin: 0 10px 20px;
      flex-grow: 1;
      display: flex;
      
      ",
                        [Function],
                        "
      
      ",
                        [Function],
                        "
      
      ",
                        [Function],
                        ";
    }
  }

  .order {
    margin: 0 0 20px;

    button {
      background: none;

      &:hover {
        cursor: pointer;
        text-decoration: underline;
      }

      span {
        color: ",
                        "#92BFB1",
                        ";
        font-weight: bold;
      }

      svg {
        position: relative;
        top: 3px;
        right: -3px;
      }
    }
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-htoDjs",
                    "target": "div",
                    "toString": [Function],
                    "usesTheme": false,
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-htoDjs eVHVmW"
                >
                  <div
                    className="order"
                  >
                    <button
                      onClick={[Function]}
                      type="button"
                    >
                      only 
                      <span>
                        sale
                      </span>
                      <Icon
                        className=""
                        color=""
                        name="down"
                        size="1em"
                      >
                        <FiChevronsDown
                          className=""
                          color=""
                          name="down"
                          size="1em"
                        >
                          <IconBase
                            attr={
                              Object {
                                "fill": "none",
                                "stroke": "currentColor",
                                "strokeLinecap": "round",
                                "strokeLinejoin": "round",
                                "strokeWidth": "2",
                                "viewBox": "0 0 24 24",
                              }
                            }
                            className=""
                            color=""
                            name="down"
                            size="1em"
                          >
                            <svg
                              color=""
                              fill="none"
                              height="1em"
                              name="down"
                              size="1em"
                              stroke="currentColor"
                              strokeLinecap="round"
                              strokeLinejoin="round"
                              strokeWidth="2"
                              style={
                                Object {
                                  "color": undefined,
                                }
                              }
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <polyline
                                key="0"
                                points="7 13 12 18 17 13"
                              />
                              <polyline
                                key="1"
                                points="7 6 12 11 17 6"
                              />
                            </svg>
                          </IconBase>
                        </FiChevronsDown>
                      </Icon>
                    </button>
                  </div>
                  <ul />
                </div>
              </StyledComponent>
            </styled.div>
          </Products>
        </Router>
      </BrowserRouter>
    </WithRouter>
  </Provider>
</WithStore>
`;
